src/utils_curl_stats.c

```diff
@@ -0,0 +1,245 @@
+{
+	CURLcode code;
+	value_t v;
+	long raw;
+
+	code = curl_easy_getinfo (curl, info, &raw);
+	if (code != CURLE_OK)
+		return -1;
+
+	v.gauge = (double)raw;
+
+	vl->values = &v;
+	vl->values_len = 1;
+
+	return plugin_dispatch_values (vl);
+} /* dispatch_size */
+
+static struct {
+	const char *name;
+	size_t offset;
+
+	int (*dispatcher)(CURL *, CURLINFO, value_list_t *);
+	const char *type;
+	CURLINFO info;
+} field_specs[] = {
+#define SPEC(name, dispatcher, type, info) \
+	{ #name, offsetof (curl_stats_t, name), dispatcher, type, info }
+
+	SPEC (total_time,              dispatch_gauge, "duration", CURLINFO_TOTAL_TIME),
+	SPEC (namelookup_time,         dispatch_gauge, "duration", CURLINFO_NAMELOOKUP_TIME),
+	SPEC (connect_time,            dispatch_gauge, "duration", CURLINFO_CONNECT_TIME),
+	SPEC (pretransfer_time,        dispatch_gauge, "duration", CURLINFO_PRETRANSFER_TIME),
+	SPEC (size_upload,             dispatch_gauge, "bytes",    CURLINFO_SIZE_UPLOAD),
+	SPEC (size_download,           dispatch_gauge, "bytes",    CURLINFO_SIZE_DOWNLOAD),
+	SPEC (speed_download,          dispatch_speed, "bitrate",  CURLINFO_SPEED_DOWNLOAD),
+	SPEC (speed_upload,            dispatch_speed, "bitrate",  CURLINFO_SPEED_UPLOAD),
+	SPEC (header_size,             dispatch_size,  "bytes",    CURLINFO_HEADER_SIZE),
+	SPEC (request_size,            dispatch_size,  "bytes",    CURLINFO_REQUEST_SIZE),
+	SPEC (content_length_download, dispatch_gauge, "bytes",    CURLINFO_CONTENT_LENGTH_DOWNLOAD),
+	SPEC (content_length_upload,   dispatch_gauge, "bytes",    CURLINFO_CONTENT_LENGTH_UPLOAD),
+	SPEC (starttransfer_time,      dispatch_gauge, "duration", CURLINFO_STARTTRANSFER_TIME),
+	SPEC (redirect_time,           dispatch_gauge, "duration", CURLINFO_REDIRECT_TIME),
+	SPEC (redirect_count,          dispatch_size,  "count",    CURLINFO_REDIRECT_COUNT),
+	SPEC (num_connects,            dispatch_size,  "count",    CURLINFO_NUM_CONNECTS),
+	SPEC (appconnect_time,         dispatch_gauge, "duration", CURLINFO_APPCONNECT_TIME),
+
+#undef SPEC
+};
+
+static void enable_field (curl_stats_t *s, size_t offset)
+{
+	*(bool *)((char *)s + offset) = true;
+} /* enable_field */
+
+static bool field_enabled (curl_stats_t *s, size_t offset)
+{
+	return *(bool *)((char *)s + offset);
+} /* field_enabled */
+
+/*
+ * Public API
+ */
+curl_stats_t *curl_stats_from_config (oconfig_item_t *ci)
+{
+	curl_stats_t *s;
+	int i;
+
+	if (ci == NULL)
+		return NULL;
+
+	s = calloc (sizeof (*s), 1);
+	if (s == NULL)
+		return NULL;
+
+	for (i = 0; i < ci->children_num; ++i)
+	{
+		oconfig_item_t *c = ci->children + i;
+		size_t field;
+
+		for (field = 0; field < STATIC_ARRAY_SIZE (field_specs); ++field)
+			if (! strcasecmp (c->key, field_specs[field].name))
+				break;
+		if (field >= STATIC_ARRAY_SIZE (field_specs))
+		{
+			ERROR ("curl stats: Unknown field name %s", c->key);
+			free (s);
+			return NULL;
+		}
+
+		if ((c->values_num != 1)
+				|| ((c->values[0].type != OCONFIG_TYPE_STRING)
```

Comment:
Can we just use `cf_util_get_boolean()` here, too? I don't see the need for allowing users to use strings as well – we don't allow this in many other cases.